include/envoy/ssl/handshaker.h

```diff
@@ -46,9 +48,38 @@ class Handshaker {
   virtual Network::PostIoAction doHandshake() PURE;
 };
 
-using HandshakerSharedPtr = std::shared_ptr<Handshaker>;
-using HandshakerFactoryCb =
-    std::function<HandshakerSharedPtr(bssl::UniquePtr<SSL>, int, HandshakeCallbacks*)>;
+/**
+ * Base interface for a combined `handshaker-and-connectioninfo` class
+ * which can both perform handshakes and provide connection-specific
+ * information.
+ */
+class HandshakerAndConnectionInfo : public Handshaker, public ConnectionInfo {
```

Comment:
I'm not thrilled with this interface, with extra virtual method. What is the goal of this PR? If we just want to provide an impl, then we might just make some mixin pattern class for ConnectionInfoImpl. If we want make the cast to ConnectionInfo safe, then perhaps make Handshaker return a ConnectionInfo pointer (in the default impl it just return `this`).

Reply:
The goal of this PR is to make the cast safe.

If we have Handshaker return a connectioninfo pointer and save that as `info_`, it's unclear to me who holds on to the SslHandshaker class for lifetime reasons. Plus `info_` also needs socketstate setter/getters, which aren't a part of either the ConnectionInfo or Handshaker interfaces.

But, I agree, I don't love this interface. WDYT?

Reply:
I don't like this either. So we're requiring one class to implement both of these interfaces. Why not just combine them into the same interface? Or if that isn't the right solution, some other refactoring appears needed.

@mattklein123 any thoughts (since you left the original comment)?

Reply:
My original thinking was to have a `connectionInfo()` method on the `Handshaker` interface and then you can handle the lifetime independently, but I don't have a strong opinion either way.